tutorial/source/ekf.ipynb

```diff
@@ -0,0 +1,247 @@
+    "for frame_num in range(1, num_frames):\n",
+    "    # sample independent process noise\n",
+    "    dx = pyro.sample('process_noise_{}'.format(frame_num), ncv.process_noise_dist(dt))\n",
+    "    xs_truth[frame_num, :] = ncv(xs_truth[frame_num-1, :], dt=dt) + dx"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, let's specify the measurements. Notice that we only measure the positions of the particle."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Measurements\n",
+    "measurements = []\n",
+    "mean = torch.zeros(2)\n",
+    "# no correlations\n",
+    "cov = 1e-5 * torch.eye(2)\n",
+    "# sample independent measurement noise\n",
+    "dzs = pyro.sample('dzs', dist.MultivariateNormal(mean, cov).expand((num_frames,)))\n",
+    "# compute measurement means\n",
+    "zs = xs_truth[:, :2] + dzs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll use a [Delta autoguide](http://docs.pyro.ai/en/0.2.1-release/contrib.autoguide.html#autodelta) to learn the position and measurement covariances. The `EKFDistribution` computes the joint log density of all of the EKF states given a tensor of sequential measurements."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def model(data):\n",
+    "    # TODO replace this with HalfNormal\n",
+    "    R = pyro.sample('pv_cov', dist.HalfCauchy(2e-6)) * torch.eye(4)\n",
+    "    Q = pyro.sample('measurement_cov', dist.HalfCauchy(1e-6)) * torch.eye(2)\n",
+    "    # observe the measurements\n",
+    "    pyro.sample('track_{}'.format(i), EKFDistribution(xs_truth[0], R, ncv,\n",
+    "                                                      Q, time_steps=num_frames),\n",
+    "                obs=data)\n",
+    "    \n",
+    "guide = AutoDelta(model)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optim = pyro.optim.Adam({'lr': 2e-2})\n",
+    "svi = SVI(model, guide, optim, loss=Trace_ELBO())\n",
+    "\n",
+    "pyro.set_rng_seed(0)\n",
+    "pyro.clear_param_store()\n",
+    "\n",
+    "for i in range(100 if not smoke_test else 2):\n",
+    "    loss = svi.step(zs)\n",
+    "    if not i % 10:\n",
+    "        print('loss: ', loss)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# retrieve states for visualization\n",
+    "R = guide()['pv_cov'] * torch.eye(4)\n",
+    "Q = guide()['measurement_cov'] * torch.eye(2)\n",
+    "ekf_dist = EKFDistribution(xs_truth[0], R, ncv, Q, time_steps=num_frames)\n",
+    "states= ekf_dist.get_states(zs)"
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {
+    "raw_mimetype": "text/html"
+   },
+   "source": [
+    "<center><figure>\n",
```

Comment:
Can you also embed the plotting code? I think it's cleaner to publish executed notebooks with cell contents so that users can fork the notebooks. If you really prefer this embedded-html idiom, then at least paste the plotting code into a non-executable code block so users can fork and re-plot.

Reply:
If the plotting code is outside of Pyro, you could still link to the package and add `import` statements in a non-executable code block.

Reply:
the reason i didnt include the plotting code is not because i prefer embedding, it's because it was really verbose formatting the data properly to use standard cognition's plotting code.  if i wanted actual reusable plotting code, i would have to rewrite most of it since pyro distributions handle the cholesky computations under the hood, etc.  the AIR tutorial also doesnt have all of its plotting code available, even on github, but i can do it if you think it significantly detracts from the tutorial.

Reply:
Ok, that sounds like a reasonable compromise.